### PR TITLE
ucarp: init at 1.5.2 / nixos/ucarp: init / nixos/test/ucarp: init

### PR DESCRIPTION
**Pull request by oxzi**
###### Motivation for this change
This pull request adds ucarp, an userspace implementation of the Common Address Redundancy Protocol (CARP).
Next to the package, a NixOS module and a test was written.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 126785` at fa5564bf run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>ucarp</li>
  </ul>
</details>

File: nixos/modules/services/networking/ucarp.nix [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+      "--upscript=${upscript}"
+      "--downscript=${downscript}"
+      "--deadratio=${toString cfg.deadratio}"
+    ] ++
+    (optional cfg.preempt "--preempt") ++
+    (optional cfg.neutral "--neutral") ++
+    (optional cfg.shutdown "--shutdown") ++
+    (optional cfg.ignoreIfState "--ignoreifstate") ++
+    (optional cfg.noMcast "--nomcast") ++
+    (optional (cfg.extraParam != null) "--xparam=${cfg.extraParam}")
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/ucarp.nix:30]:
```suggestion
    ] ++ (optional cfg.preempt "--preempt")
    ++ (optional cfg.neutral "--neutral")
    ++ (optional cfg.shutdown "--shutdown")
    ++ (optional cfg.ignoreIfState "--ignoreifstate")
    ++ (optional cfg.noMcast "--nomcast")
    ++ (optional (cfg.extraParam != null) "--xparam=${cfg.extraParam}")
```

File: pkgs/servers/ucarp/default.nix [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+stdenv.mkDerivation rec {
+  pname = "ucarp";
+  version = "1.5.2";
+
+  src = fetchurl {
+    url = "https://download.pureftpd.org/pub/ucarp/ucarp-${version}.tar.bz2";
+    sha256 = "0qidz5sr55nxlmnl8kcbjsrff2j97b44h9l1dmhvvjl46iji7q7j";
+  };
+
+  nativeBuildInputs = [ libpcap ];
```
Comment (SuperSandro2000) [inline pkgs/servers/ucarp/default.nix:12]:
```suggestion
  buildInputs = [ libpcap ];
```

**Comment by oxzi** [comment]
Thanks @SuperSandro2000, I have applied your suggested changes. Would you please take a second look?

**Comment by markuskowa** [comment]
The sources are from 2010. Is it really a good idea to integrate unmaintained software, specially when it aims high availability solutions?

**Comment by oxzi** [comment]
> The sources are from 2010. Is it really a good idea to integrate unmaintained software, specially when it aims high availability solutions?

Thanks for bringing up this issue, @markuskowa.

Prior to packaging ucarp and finalizing this PR, I took a look at other distributions.
The following list is a result from the more famous distributions shipping ucarp, fetched by [Repology](https://repology.org/project/ucarp/versions).

- Alpine Linux uses the same tarball, [build script](https://git.alpinelinux.org/aports/tree/main/ucarp/APKBUILD).
- Arch Linux uses the same tarball, [build script](https://github.com/archlinux/svntogit-community/blob/packages/ucarp/trunk/PKGBUILD).
- Debian uses the same version within a [Debianified repository](https://salsa.debian.org/eevans/ucarp/-/blob/master/debian/changelog).
- DragonFly BSD uses an abandoned fork <https://github.com/jedisct1/UCarp>, [package description](https://github.com/DragonFlyBSD/DPorts/blob/master/net/ucarp/pkg-descr).
- Fedora uses the same tarball, [build script](https://src.fedoraproject.org/rpms/ucarp/blob/main/f/ucarp.spec).
- FreeBSD uses the same fork as DragonFly BSD, [package description](https://cgit.freebsd.org/ports/tree/net/ucarp/pkg-descr).
- MacPorts uses the same tarball or at least domain, [build script](https://github.com/macports/macports-ports/blob/master/net/ucarp/Portfile).
- Ubuntu packages the same version, [information](https://launchpad.net/ubuntu/+source/ucarp/).

In a nutshell, most major distributions provide the same software.

Even if the software is old, there are no known security vulnerabilities.
Furthermore, one can find quite recent blog posts about ucarp online.
For the best CARP experience, one should probably use OpenBSD, but ucarp works and is actively used.

For myself, I had a reason to use it and package it for NixOS.
This should be enough as an argument.

File: nixos/modules/services/networking/ucarp.nix [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    ++ (optional cfg.noMcast "--nomcast")
+    ++ (optional (cfg.extraParam != null) "--xparam=${cfg.extraParam}")
+  );
+in {
+  options.networking.ucarp = {
+    enable = mkEnableOption "ucarp, userspace implementation of CARP";
+
+    interface = mkOption {
+      type = types.str;
+      description = "Network interface to bind to";
```
Comment (aanderse) [inline nixos/modules/services/networking/ucarp.nix:38]: `description` fields in `mkOption` should end with a period.

File: nixos/modules/services/networking/ucarp.nix [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    };
+
+    srcIp = mkOption {
+      type = types.str;
+      description = "Source (real) IP address of this host";
+    };
+
+    vhId = mkOption {
+      type = types.ints.between 1 255;
+      description = "Virtual IP identifier (1-255)";
```
Comment (aanderse) [inline nixos/modules/services/networking/ucarp.nix:49]: The "(1-255)" is redundant because documentation will render that in the type description.

File: nixos/modules/services/networking/ucarp.nix [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+      example = 1;
+    };
+
+    pass = mkOption {
```
Comment (aanderse) [inline nixos/modules/services/networking/ucarp.nix:53]: 😢

File: nixos/modules/services/networking/ucarp.nix [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+      default = 1;
+    };
+
+    advSkew = mkOption {
+      type = types.ints.unsigned;
+      description = "Advertisement skew in seconds";
+      default = 0;
+    };
+
+    upscript = mkOption {
```
Comment (aanderse) [inline nixos/modules/services/networking/ucarp.nix:87]: You might consider making this `type` a `path` and then using `writeShellScript` (or something like that) in the `example`. Even if you don't drink enough nix kool-aid to buy into "`/bin/sh` is impure and impure is bad" argument this module is going to be part of nixpkgs and should probably promote these ideals.

**Review by aanderse — COMMENTED** [review]
Quick overview of the module without looking into the software being packaged... LGTM :+1:

I left a few minor nitpicks that are cosmetic, but I would appreciate if you could make those changes before this PR is merged.

**Comment by markuskowa** [comment]
> Prior to packaging ucarp and finalizing this PR, I took a look at other distributions.
> The following list is a result from the more famous distributions shipping ucarp, fetched by Repology.

OK, that is a valid argument. However, I recommend adding a note to package description and module description, stating that the project is unmaintained.

File: nixos/modules/services/networking/ucarp.nix (hunk 3 above)
Comment (oxzi) [reply nixos/modules/services/networking/ucarp.nix]: Changed for all `mkOptions`.

File: nixos/modules/services/networking/ucarp.nix (hunk 5 above)
Comment (oxzi) [reply nixos/modules/services/networking/ucarp.nix:51]: Added an additional `passwordFile` option. I even planed this initially, but totally forgot about it. Thanks for the reminder.

File: nixos/modules/services/networking/ucarp.nix (hunk 6 above)
Comment (oxzi) [reply nixos/modules/services/networking/ucarp.nix:105]: Done.

**Comment by oxzi** [comment]
Thanks for your review, @aanderse.

I updated this PR once again and tried to address the issues raised by @aanderse and @markuskowa.
- Both package as well as NixOS module have a note attached regarding the potential unmaintained state of the ucarp software.
- The `mkOption` descriptions are ending with a dot.
- The redundant type hint for `vhId` was removed.
- A second `passwordFile` option was added which is mutual exclusive to `pass`.
- Both `upscript` and `downscript` are now a `path`.

File: nixos/modules/services/networking/ucarp.nix [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+      description = ''
+        Shared password between CARP hosts.
+
+        Either <literal>pass</literal> or <literal>passwordFile</literal> must
+        be set.
+      '';
+      default = null;
+    };
+
+    passwordFile = mkOption {
```
Comment (aanderse) [inline nixos/modules/services/networking/ucarp.nix:62]: I suggest we do not make it easy for people. If someone really wants to shoot themselves in the foot, or has a legitimate use case of why it is OK to store a password in the nix store, they can always do this:
```
networking.ucarp.passwordFile = pkgs.writeText "badidea.txt" "password-everyone-on-this-server-can-read";
```
```

**Review by aanderse — CHANGES_REQUESTED** [review]
Great work! I think this is pretty much good to go... with one minor nag on my side: the `pass` option. I really want to emphasize that a goal of NixOS is to remove **every** password option, so I don't like merging PRs which introduce new ones...

This software seems to have a password file option built right in! That is **great**. We should really encourage that. Of course we don't want to tell our users they **can't** do something insecure, just that they shouldn't - hence only having `passwordFile` allows people to use `pkgs.writeText` and store their password in the nix store if they need to.

I approve this module and want to say you did a great job on it. If you're dead set against removing the `pass` option please state your reasons why and we can discuss.

**Comment by oxzi** [comment]
I am completely with you regarding keeping secrets out of the nix store, @aanderse.

Thus, I removed the `pass` option and exclusion logic between `pass` and `passwordFile`.
This way the code is smaller and no password lands in the nix store, which was never designed to hold secrets anyway.

Could you please take another look? Thanks.

**Review by aanderse — APPROVED** [review]
Module looks great and comes with a test ❤️. Full bonus points awarded! :tada:

Great job.

**Review by markuskowa — APPROVED** [review]
All clear from my side.

**Comment by oxzi** [comment]
Once again, I wanted to thank you for the excellent and quick reviews. Thanks a million for your work, @SuperSandro2000, @markuskowa, and @aanderse.
